src/components/NotesMindMap/NotesMindMapReducer.js

```diff
@@ -14,114 +14,128 @@ import {
-          isNote: child.isNote,
-          group: 'children',
-        });
+        nodes = addNoteToGraph(nodes, child);
         edges.push({ from: child.parent.id, to: child.id });
       });
-      notes = addGroupTagToNote(notes, state.selectedNote.id);
+      nodes = addGroupTagToNodes(nodes, edges);
     }
-    return { ...state, notes, edges };
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleEditNoteNameAction = () => {
     const noteToEdit = data;
-    return { ...state, showNoteNameEditor: true, selectedNote: noteToEdit };
+    return {
+      ...state,
+      showNoteNameEditor: true,
+      selectedNote: noteToEdit,
+    };
   };
 
   const handleNoteNameUpdateRequestSuccessAction = () => {
     const updatedNote = data;
-    let notes = [...state.notes];
-    notes = notes.filter(note => note.id !== updatedNote.id);
-    notes.push({
-      id: updatedNote.id,
-      label: updatedNote.name,
-      name: updatedNote.name,
-      isNote: updatedNote.isNote,
-    });
-    return { ...state, notes };
+    let nodes = [...state.nodes];
+    nodes = nodes.filter(note => note.id !== updatedNote.id);
+    return {
+      ...state,
+      nodes: addNoteToGraph(nodes, updatedNote),
+    };
   };
 
   const handleCreateNoteSuccessAction = () => {
     const newNote = data;
-    let notes = [...state.notes];
+    let nodes = [...state.nodes];
     let edges = [...state.edges];
-    notes.push({
-      id: newNote.id,
-      name: newNote.name,
-      label: newNote.name,
-      isNote: newNote.isNote,
-      group: 'children',
-    });
     edges.push({ from: newNote.parent.id, to: newNote.id });
-    notes = addGroupTagToNote(notes, state.selectedNote.id);
-    return { ...state, notes, edges };
+    nodes = addGroupTagToNodes(addNoteToGraph(nodes, newNote), edges);
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleDeleteNoteRequestSuccessAction = () => {
     const noteToDelete = data;
-    let notes = [...state.notes];
-    let edges = [...state.edges];
-    let graph = removeNoteFromGraph(notes, edges, noteToDelete);
-    notes = graph.notes;
-    edges = graph.edges;
-    return { ...state, notes, edges };
+    const { nodes, edges } = removeNodeFromGraph(
+      state.nodes,
+      state.edges,
+      noteToDelete,
+    );
+    return {
+      ...state,
+      nodes,
+      edges,
+    };
   };
 
   const handleChangeParentNoteRequestSuccess = () => {
     const noteId = data.noteId;
     const newParentId = data.newParentId;
     let edges = [...state.edges];
-    let notes = [...state.notes];
+    let nodes = [...state.nodes];
     const oldParent = edges.find(edge => edge.to === noteId);
     edges = edges.filter(edge => edge.to !== noteId);
     edges.push({ from: newParentId, to: noteId });
 
-    notes = updateGroupOfOldParent(notes, edges, newParentId, oldParent.from);
+    nodes = updateGroupOfOldParent(nodes, edges, newParentId, oldParent.from);
```

Comment:
If a node has children, its group should be "parent", otherwise its group should be "children".

Why do we care if this node is an old parent? We have the `oldParent` node. Just use its id to see in `edges` if it has any children or not. Why in addition to the `oldParent.id` we need to pass the `newParentId` ?

Reply:
Done